### PR TITLE
Add some explanations for the strategy solving the Poisson problems with non-homogeneous Dirichlet boundary conditions

### PR DESCRIPTION
**Pull request by wei3li**

### CODE REVIEW

**Comment by amartinhuertas** [comment]
Hi @wei3li ! Thanks for your work.

**Comment by amartinhuertas** [comment]
I have generated the page of the tutorial locally on my machine, and I noted that there are some typos, and line breaks missing  in the result. This is in particular what I see:

Screenshot from 2022-01-27 10-41-19

I guess that you modified and pushed the tutorial sources without double checking that the generated page looks as you were expecting, right? No problem. Indeed, I couldnt find the instructions on the README file nor in the documentation of the Tutorials repo in order to generate locally the HTML pages of the tutorials.

To this end, you have to follow the following instructions once:

```
julia --project=docs   # From the Unix shell, located at the root of Tutorials repo 
develop .              # From the Julia package manager prompt
instantiate            # "" 
build                  # "" 
exit()                 # From the Julia REPL
```

and then, each time that you perform a change on the tutorial sources, you have to execute the following from the shell:

```
julia --project=docs docs/make.jl # From the Unix shell, located at the root of Tutorials repo 
```

to generate the HTML pages. The pages generated are available at "Tutorials/docs/build/pages". You can open them, e.g., with Google Chrome.  Perhaps, once you are familiar with these instructions, we may add a section to the README file, something like "Generating tutorial HTML pages locally"  or in the Wiki of the repo, https://github.com/gridap/Tutorials/wiki.

As a first step in the review process, can you please modify the sources of the tutorial so that the generated page looks as you were expecting and push your changes? Thanks! 

On the other hand, with respect to the contents, while they are correct, they go may be too straight to the point. I think is better to add some more narrative and motivation about why we are doing this. I can do this part once you add your changes, ok?

**Comment by amartinhuertas** [comment]
For the records, @wei3li was able to generate locally the ipynb files using Literate.jl, and they looked ok for the changes at hand. The issue seems to be in the generated HTML pages, nor actually in the ipynb files.

**Comment by amartinhuertas** [comment]
Hereby I confirm that, after the changes in 84e6616, the issues with the HTML file have disappeared. Nice work @wei3li ! I will know continue with my review.

File: src/poisson_dev_fe.jl [hunk 1]
```diff
@@ -571,6 +571,29 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
 
 # ## A low-level implementation of the residual integration and assembly
 
+# The discretized variational form of the Poisson problem reads as:
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:574]: I would start here with some introductory text towards an smoother transition + some motivation. Something like (did not check that renders ok, might need edits, etc.):

```
In the rest of the tutorial we aim to solve a Poisson equation with homogeneous source term, i.e.,  `$f=0$` 
and non-homogeneous Dirichlet boundary conditions $u=g_0$ on $\Gamma_D$, with  $\Gamma_D$ being
 the whole boundary of the box. While the strong imposition of non-homogeneous Dirichlet boundary 
conditions in Gridap is done under the hood by modifying the global assembly process by subtracting 
the contributions of boundary conditions to the right hand side of the linear system, in this tutorial,
 for simplicity, we follow a different approach. This approach requires: (1) to be able to assemble the 
residual of the PDE for an arbitrary finite element function $\hat{u}_h$ (current section). (2) to be able 
to assemble the coefficient matrix of the finite element linear system (next section). We briefly outline 
this solution approach in the sequel. 
```
What do you think?

File: src/poisson_dev_fe.jl [hunk 2]
```diff
@@ -571,6 +571,29 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+# >Find $u_h \in V_h^{\Gamma_D}=\{v_h \in V_h:v_h=g_0 \; on \; \Gamma_D\}$ such that
+# >```math
+# >(u_h,v_h)=l(v_h)\quad \forall v_h \in V_h^0
+# >```
+# >where $V_h^0=\{v_h \in V_h:v_h=0 \; on \; \Gamma_D\}$.
+#
+# Take $u_h = \hat{u}_h-w_h$ where $w_h \in V_h^0$ and $\hat{u}_h \in V_h^{\Gamma_D}$, then
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:581]: I would rewrite this sentence as follows. 

```
If we take an arbitrary function $\hat{u}_h in V_h^{\Gamma_D}$, and 
compute $w_h \in V_h^0$ as the solution of the following discrete variational problem: 

Then, it is easy to see that `$u_h = \hat{u}_h-w_h$`. This is the strategy 
that we implement in the sequel. Note that, in a previous section, 
we created a FE function using `rand` for initializing the values of the 
degrees of freedom. This function indeed plays the role of \hat{u}_h.
```
What do you think?

File: src/poisson_dev_fe.jl [hunk 3]
```diff
@@ -571,6 +571,29 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+# ```math
+# a(w_h,v_h)=a(\hat{u}_h,v_h)-l(v_h)\quad \forall v_h,w_h \in V_h^0.
+# ```
+#
+# Find some $\hat{u}_h$, solve that eqation for $w_h$, then the solution is
+# ```math
+# u_h=\hat{u}_h-w_h.
+# ```
+#
+# In the integration and assembly process, we first use the `rand` function to assign random values to the free DoFs to form $\hat{u}_h$ (name the randomly fixed DoFs as $\vec{\hat{u}}$). In this particular problem, $l(v_h)$ is obviously zero, so just compute $a(\hat{u}_h,v_h)$ for each element, then assemble the results to obtain the global force vector $\vec{b}$. On the left hand side, compute every element-wise $a(w_h,v_h)$, and assemble them to form global stiffness matrix $\textbf{A}$. Solve the linear system $\textbf{A}\vec{x}=\vec{b}$ for $\vec{x}$, then the final solution for the free DoFs is
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:591]: I think that this part can be removed after the modifications above. We already made clear that we need to assembly the residual and the matrix, and how we initialized the FE function. Agreed?

**Review by amartinhuertas — CHANGES_REQUESTED** [review]
Hi @wei3li, I have finished my review. You can proceed with the requested changes. Thanks!

File: src/poisson_dev_fe.jl [hunk 4]
```diff
@@ -571,6 +571,29 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+# >(u_h,v_h)=l(v_h)\quad \forall v_h \in V_h^0
+# >```
+# >where $V_h^0=\{v_h \in V_h:v_h=0 \; on \; \Gamma_D\}$.
+#
+# Take $u_h = \hat{u}_h-w_h$ where $w_h \in V_h^0$ and $\hat{u}_h \in V_h^{\Gamma_D}$, then
+# ```math
+# a(w_h,v_h)=a(\hat{u}_h,v_h)-l(v_h)\quad \forall v_h,w_h \in V_h^0.
+# ```
+#
+# Find some $\hat{u}_h$, solve that eqation for $w_h$, then the solution is
```
[the remark this answers is not in the document]
Comment (wei3li) [reply src/poisson_dev_fe.jl:586]: I found a VSCode plugin for spell checking. Previously I was using JetBrain IDEs, which come with spell checking functionality out of the box. I thought VSCode is very similar, sorry about that!

File: src/poisson_dev_fe.jl (hunk 1 above)
Comment (wei3li) [reply src/poisson_dev_fe.jl:574]: The background introduction is definitely needed.

File: src/poisson_dev_fe.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (amartinhuertas) [reply src/poisson_dev_fe.jl:586]: No worries

[the remark this answers is not in the document]
Comment (amartinhuertas) [reply src/poisson_dev_fe.jl:586]: For my knowledge, how is the plugin called? I may be using this as well in the future.

[the remark this answers is not in the document]
Comment (wei3li) [reply src/poisson_dev_fe.jl:586]: Code Spell Checker

File: src/poisson_dev_fe.jl (hunk 3 above)
Comment (wei3li) [reply src/poisson_dev_fe.jl]: agree with you.

File: src/poisson_dev_fe.jl (hunk 2 above)
Comment (wei3li) [reply src/poisson_dev_fe.jl]: definitely better than mine

File: src/poisson_dev_fe.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (amartinhuertas) [reply src/poisson_dev_fe.jl]: Ok, thanks. I have installed it and seems to work as a charm!

File: src/poisson_dev_fe.jl [hunk 5]
```diff
@@ -571,6 +571,21 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
 
 # ## A low-level implementation of the residual integration and assembly
 
+# In the rest of the tutorial we aim to solve a Poisson equation with homogeneous source term, i.e., $f=0$ and non-homogeneous Dirichlet boundary conditions $u=g_0$ on $\Gamma_D$, with  $\Gamma_D$ being the whole boundary of the model. While the strong imposition of non-homogeneous Dirichlet boundary conditions in Gridap is done under the hood by modifying the global assembly process by subtracting the contributions of boundary conditions to the right hand side of the linear system, in this tutorial, for simplicity, we follow a different approach. This approach requires: (1) to be able to assemble the residual of the PDE for an arbitrary finite element function $\hat{u}_h$ (current section); (2) to be able to assemble the coefficient matrix of the finite element linear system (next section). We briefly outline this solution approach in the sequel.
+
+# The discretized variational form of the Poisson problem reads as:
+# >Find $u_h \in V_h^{\Gamma_D}=\{v_h \in V_h:v_h=g_0 \; on \; \Gamma_D\}$ such that
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:577]: There is a subtle, but important mathematical detail in here. At the continuous problem level, we can enforce `v=g_0`. However, this is not actually the case at the discrete problem level, as, generally,  `g_0` might not be in the FE space. What we do in practice is to enforce `v_h=\Pi_h(g_0)` where $\Pi_h$ denotes a suitable projection operator, for example, the interpolation operator. Can you modify the definition of the FE space to reflect this subtletie? Thanks!

File: src/poisson_dev_fe.jl [hunk 6]
```diff
@@ -571,6 +571,21 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+# >```math
+# >a(u_h,v_h)=l(v_h)\quad \forall v_h \in V_h^0
+# >```
+# >where $V_h^0=\{v_h \in V_h:v_h=0 \; on \; \Gamma_D\}$.
+#
+# If we take an arbitrary function $\hat{u}_h \in V_h^{\Gamma_D}$, and compute $w_h \in V_h^0$ as the solution of the following discrete variational problem:
+# ```math
+# a(w_h,v_h)=a(\hat{u}_h,v_h)-l(v_h)\quad \forall v_h,w_h \in V_h^0.
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:585]: replace . by , at the end of the sentence

File: src/poisson_dev_fe.jl [hunk 7]
```diff
@@ -571,6 +571,21 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+# ```
+# Then, it is easy to see that $\mathbf{ u_h=\hat{u}_h-w_h }$. This is the strategy that we implement in the sequel. Note that, in previous section, we created a FE function using `rand` for initializing the values of the degrees of freedom. This function indeed plays the role of $\hat{u}_h$.
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:587]: replace Then by then

File: src/poisson_dev_fe.jl [hunk 8]
```diff
@@ -571,6 +571,21 @@ low_level_manual_gradient_dv_array = lazy_map(Broadcasting(Operation(⋅)),inv_J
...
+
```
Comment (amartinhuertas) [inline src/poisson_dev_fe.jl:588]: Last comment. Replace the following sentence near the end of the tutorial

# Now we can obtain the free DOFs and add the solution to the initial guess

by 

# Now we can obtain the free DOFs by subtracting the solution to the initial guess

There was a typo in the tutorial. Agreed?

**Review by amartinhuertas — COMMENTED** [review]
@wei3li ... thanks for your work! I just added a pair of additional comments. The tutorial is looking much better!

File: src/poisson_dev_fe.jl (hunk 8 above)
Comment (wei3li) [reply src/poisson_dev_fe.jl:588]: Agree.
"subtract ... from ..." is more appropriate.

File: src/poisson_dev_fe.jl (hunk 5 above)
Comment (wei3li) [reply src/poisson_dev_fe.jl:577]: Yes, that's important to notice. I prefer to use $g_0^h$ to denote the discretized version of function $g_0$.

Comment (amartinhuertas) [reply src/poisson_dev_fe.jl]: ok, let us use $g_0^h$, and add a comment to briefly define how it is defined.

File: src/poisson_dev_fe.jl (hunk 8 above)
Comment (amartinhuertas) [reply src/poisson_dev_fe.jl:588]: Definitely! from instead of to

**Review by amartinhuertas — APPROVED** [review]
@wei3li ... I am satisfied with the current changes. I will accept the PR now. Thanks!

**Comment by wei3li** [comment]
One more commit, I added the definition to $g_0^h$.

**Comment by amartinhuertas** [comment]
> One more commit, I added the definition to $g_0^h$.

Ok, I already noticed it. The changes are already available at the official tutorials repo.

**Pull request merged**